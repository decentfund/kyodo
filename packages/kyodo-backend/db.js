const mongoose = require('mongoose');
const { SystemError } = require('./errors');

const Schema = mongoose.Schema;

const Period = require('./models/Period');
const User = require('./models/User');
const Tip = require('./models/Tip');

mongoose.connection.on('error', e => {
  if (e.message.code === 'ETIMEDOUT') {
    console.log(e);
  }
  console.log(e);
});

mongoose.connection.once('open', () => {
  console.log(`MongoDB successfully connected to mongoUri`);
});

const initDb = () => mongoose.connect('mongodb://localhost/colony');

const taskSchema = new mongoose.Schema({
  taskId: String,
  taskTitle: String,
  specificationHash: String,
  finalized: Boolean,
  cancelled: Boolean,
  dateCreated: Date,
  dueDate: Date,
  potId: Number,
  // domainId: Number,
  // id: Number,
  // skillId: Number
});

const colonySchema = new mongoose.Schema({
  colonyName: String,
  colonyId: String,
  colonyAddress: String,
  tokenAddress: String,
  tokenName: String,
  tokenSymbol: String,
  creationBlockNumber: Number,
  creationDate: Date,
  periodIds: Array,
});

colonySchema.virtual('currentPeriodId').get(function() {
  return this.periodIds[this.periodIds.length - 1];
});

const domainSchema = new Schema({
  domainId: Number,
  domainTitle: String,
  localSkillId: Number,
  potId: Number,
});

const Task = mongoose.model('Task', taskSchema);
const Colony = mongoose.model('Colony', colonySchema);
const Domain = mongoose.model('Domain', domainSchema);

const getColonyById = async colonyId => {
  return await Colony.findOne({ colonyId });
};

const getCurrentUserPeriod = async (alias, periodId) => {
  const user = await User.findOne({ alias });
  const period = await Period.findOne({ user: user._id, periodId });
  const sentTips = await Tip.find({ from: user, periodId });
  const usedPoints = sentTips.reduce((sum, v) => sum + v.amount, 0);

  if (!period) throw new SystemError('Cannot find user current period');

  return {
    ...period,
    initialBalance: period.balance,
    balance: period.balance - usedPoints,
  };
};

const getDomainByCode = async code => {
  const domain = await Domain.findOne({ domainTitle: code });
  if (!domain) throw Error('Domain not found');

  return domain;
};

module.exports = {
  initDb,
  Task,
  Colony,
  Domain,
  Tip,
  User,
  Period,
  getColonyById,
  getCurrentUserPeriod,
  getDomainByCode,
};
