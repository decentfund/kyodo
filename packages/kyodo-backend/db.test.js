import 'babel-polyfill';
import { Period, Colony, getCurrentUserPeriod } from './db';
import {
  connectMongoose,
  clearDb,
  disconnectMongoose,
  createColony,
  createUser,
} from './test/helpers';

beforeAll(connectMongoose);
beforeEach(clearDb);
afterAll(disconnectMongoose);

describe('db', () => {
  beforeEach(createColony);

  describe('colony', () => {
    it('returns proper current period id', async () => {
      const colony = await Colony.findOne();
      expect(colony.currentPeriodId).toBe(0);
    });

    it('returns proper current period id', async () => {
      const colony = await Colony.findOne();
      colony.periodIds.push(0);
      await colony.save();
      expect(colony.currentPeriodId).toBe(0);
    });
  });
});

describe('getCurrentUserPeriod', async () => {
  let user;

  beforeEach(async () => {
    user = await createUser();
  });

  it('throws error if user period is not found', async () => {
    try {
      await getCurrentUserPeriod('user', 0);
    } catch (e) {
      expect(e).toMatchSnapshot();
    }
  });

  it('returns period correctly', async () => {
    // creating new user period
    const period = new Period({ user, periodId: 0, balance: 0, tips: [] });
    await period.save();

    const periodResponse = await getCurrentUserPeriod('user', 0);
    expect(periodResponse.initialBalance).toBe(0);
    expect(periodResponse.balance).toBe(0);
  });
});
