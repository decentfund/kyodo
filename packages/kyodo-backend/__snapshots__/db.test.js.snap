// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getCurrentUserPeriod throws error if user period is not found 1`] = `[SystemError: Cannot find user current period]`;
